# Aufgabenblatt 6: Pandas

importiere die folgenden Bibliotheken
- pandas
- numpy
- random
- string

In [1296]:
import pandas as pd
import numpy as np
import random
import string

from tornado.escape import utf8

## Aufgabe 1: Erstelle einen eigenen Dataframe
### a) Dataframe erstellen mit manueller Eingabe

erstelle einen DataFrame `data_1a` mit den Kennzahlen:
- Column `column_string_1` mit *'String_1', 'String_2', 'Stirng_3'*
- Column `column_mix` mit *'rot', 'rot', 'blau'*
- Column `column_int` mit *1, 2, 2002*
- Column `column_float` mit *1.5, 1.7, 1.9*

In [1297]:
data_1a = pd.DataFrame({
    "column_string_1": ["String_1", "String_2", "String_3"],
    "column_mix": ["rot", "rot", "blau"],
    "column_int": [1, 2, 2002],
    "column_float": [1.5, 1.7, 1.9]
})
data_1a

,column_string_1,column_mix,column_int,column_float
0,String_1,rot,1,1.5
1,String_2,rot,2,1.7
2,String_3,blau,2002,1.9


### b) Dateframe erstellen mit zufälligen und automatisch hinzugefügten Werten

Erstelle einen Dataframe `data_1b`
- der Dataframe hat `4` Columns mit jeweils `100` rows
- der Dataframe hat die Column `column_string` mit *2 zufälligen Buchstaben pro Eintrag*
- der Dataframe hat die Column `column_string_2` mit *gebe zufällig den Werten **rot** und **Blau** und 
    - der Wert rot soll 75% der Ausprägungen haben*
- der Dataframe hat die Column `column_int` mit *zufälligen Werte im Bereich von 1 bis 10*
- der Dataframe hat die Column `column_float` mit *zufällige Werte zwischen -100.00 und 100.00*
<br> <br>*Tipp:* 
- random.choice, string.ascii_lowercase, uniform

In [1298]:
data_1b = pd.DataFrame({
    "column_string": [''.join(random.choices(string.ascii_letters, k=2)) for _ in range(100)],
    "column_string_2": [random.choice(["rot", "rot", "rot", "blau"]) for _ in range(100)],
    "column_int": [random.randint(1, 10) for i in range(100)],
    "column_float": [random.uniform(-100, 100) for _ in range(100)],
})
data_1b

,column_string,column_string_2,column_int,column_float
0,Sx,rot,4,-57.280156
1,Tj,rot,3,-67.094887
2,uK,rot,10,-94.073891
3,yb,blau,1,-54.456918
4,Ji,rot,3,-7.340326
...,...,...,...,...
95,Zv,blau,5,-85.278077
96,qN,blau,10,-55.239325
97,bJ,rot,1,91.435313
98,TY,rot,6,24.407883


## Aufgabe 2: verbinden von DataFrames

führe die beiden Dataframes `data_1a` und `data_1c` (alle Einträge) zusammen:
- die join/merge Parameter Column ist gleich der Column int `column_int`,
- verwende die Funktion `merge` und verwende für die Übungszwecke die verschiedenen Parameter:
    1. gebe an aus welchem Dataframe kommt der Eintrag
    2. wenn Columnnamen in beiden Dataframes vorkommen, sollen diese mit rechts und links erweitert werden

<br> 101 rows × 8 columns

In [1299]:
data_2 = pd.merge(
    data_1a,
    data_1b,
    on="column_int",
    how="outer",
    indicator=True,
    suffixes=("_A", "_B")
)

data_2


,column_string_1,column_mix,column_int,column_float_A,column_string,column_string_2,column_float_B,_merge
0,String_1,rot,1,1.5,yb,blau,-54.456918,both
1,String_1,rot,1,1.5,sF,rot,82.168786,both
2,String_1,rot,1,1.5,of,rot,12.496530,both
3,String_1,rot,1,1.5,rg,blau,61.379347,both
4,String_1,rot,1,1.5,EZ,rot,33.792145,both
...,...,...,...,...,...,...,...,...
96,NaN,NaN,10,NaN,of,rot,25.937956,right_only
97,NaN,NaN,10,NaN,JC,rot,83.944678,right_only
98,NaN,NaN,10,NaN,zz,rot,38.914995,right_only
99,NaN,NaN,10,NaN,qN,blau,-55.239325,right_only


## Aufgabe 3: Datenimport  und Datenaufbereitung
### a) import von csv und excel Datein
- importiere die csv Datei `2023`
- bestimme die folgenden Parameter, sodass die Datei richtig gelesen werden kann:
    - header
    - seperator
    - encoding

In [1300]:
df_2023 = pd.read_csv(
    filepath_or_buffer="ressources/2023.csv",
    header=0,
    sep="\t",
    encoding="latin1"
)
df_2023

# \t für Tab als Seperator | Kein plan warum utf-8 (Standard) nicht funktioniert -> deshalb latin1


,Unnamed: 0,Getränk,Region,Bestellnummer,Rechnungsnummer,Kundennummer,Listenpreis,Rabatt in Prozent,Verkaufspreis,Rabatt Preis,Einkaufpreis,VK Peris in USD
0,205,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,n96.71,131.76
1,8920,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,n132.81,-6.64,99.61,135.71
2,4116,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,1716,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.1,185.14
4,10901,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...,...
10922,5585,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.0,87.49
10923,6251,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,10719,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,6407,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- importiere alle sheets aus der Excel_Datei Datei `before_2023` und speicher diese in einzelne DatFrames

In [1301]:
df_2020 = pd.read_excel("ressources/before_2023.xlsx", sheet_name="2020")
df_2021 = pd.read_excel("ressources/before_2023.xlsx", sheet_name="2021")
df_2022 = pd.read_excel("ressources/before_2023.xlsx", sheet_name="2022")

df_2020.info()
df_2021.info()
df_2022.info()



<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         2000 non-null   int64  
 1   Getränk            2000 non-null   str    
 2   Region             2000 non-null   str    
 3   Bestellnummer      2000 non-null   int64  
 4   Rechnungsnummer    2000 non-null   str    
 5   Kundennummer       2000 non-null   int64  
 6   Listenpreis        2000 non-null   float64
 7   Rabatt in Prozent  2000 non-null   int64  
 8   Verkaufspreis      2000 non-null   float64
 9   Rabatt Preis       2000 non-null   float64
 10  Einkaufpreis       2000 non-null   float64
 11  VK Peris in USD    2000 non-null   float64
dtypes: float64(5), int64(4), str(3)
memory usage: 187.6 KB
<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             ----

### b) Datenaufbereitung des Dataframes `2023`
- schaue dir den neuen Datensatz an? 
-   beschreibe die Grösse des Datensatz (rows/records und columns)

In [1302]:
df_2023.shape

(10927, 12)

- bestimme die mathematischen Werte wie z.B. `count`, `mmean`,`max`, `min`, `std`, `25%`, `50%` und `75%` des Datensatzes

In [1303]:
df_2023.describe()

,Unnamed: 0,Bestellnummer,Kundennummer,Listenpreis,Rabatt in Prozent,Rabatt Preis,VK Peris in USD
count,10927.00000,10919.000000,10926.000000,10914.000000,10914.000000,10926.000000,10926.000000
mean,5463.00000,5032.738896,149.100037,155.527156,1.273594,-2.014615,165.138933
std,3154.49753,2906.072110,85.848563,43.382146,2.698154,4.491946,46.296567
min,0.00000,0.000000,1.000000,80.020000,0.000000,-44.540000,71.000000
25%,2731.50000,2516.500000,75.000000,118.160000,0.000000,0.000000,125.230000
50%,5463.00000,5033.000000,149.000000,156.480000,0.000000,0.000000,165.845000
75%,8194.50000,7548.500000,223.000000,192.832500,0.000000,0.000000,204.845000
max,10926.00000,10079.000000,299.000000,246.620000,10.000000,0.000000,264.780000


- erhalte die allgemeinen Informationen zum Datensatz wie z.B. Null-Werte, Datentypen, Columnnamen

In [1304]:
df_2023.info()

<class 'pandas.DataFrame'>
RangeIndex: 10927 entries, 0 to 10926
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         10927 non-null  int64  
 1   Getränk            10912 non-null  str    
 2   Region             10924 non-null  str    
 3   Bestellnummer      10919 non-null  float64
 4   Rechnungsnummer    10918 non-null  str    
 5   Kundennummer       10926 non-null  float64
 6   Listenpreis        10914 non-null  float64
 7   Rabatt in Prozent  10914 non-null  float64
 8   Verkaufspreis      10914 non-null  str    
 9   Rabatt Preis       10926 non-null  float64
 10  Einkaufpreis       10899 non-null  str    
 11  VK Peris in USD    10926 non-null  float64
dtypes: float64(6), int64(1), str(5)
memory usage: 1.0 MB


- schaue dir nur die esten 5 Einträge an

In [1305]:
df_2023.head(5)

,Unnamed: 0,Getränk,Region,Bestellnummer,Rechnungsnummer,Kundennummer,Listenpreis,Rabatt in Prozent,Verkaufspreis,Rabatt Preis,Einkaufpreis,VK Peris in USD
0,205,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,n96.71,131.76
1,8920,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,n132.81,-6.64,99.61,135.71
2,4116,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,1716,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.1,185.14
4,10901,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46


- die Column `Verkaufspreis` und `Einkaufpreis` sollen nur aus Zahlen bestehen und den Datentyp **float** besitzen
- entferne die Buchstaben und erstelle einen neuen Dataframe mit Float Werten in den beiden Columns
    - Fehler im Datenframe drüber sind: 
        - `Verkaufspreis` - row 1
        - `Einkaufspreis` - row 0
<br><br> **Tipp** manuelle Lösung mittels replace z.B.

In [1306]:
df_2023_cleaned = df_2023.copy()
df_2023_cleaned["Verkaufspreis"] = df_2023_cleaned["Verkaufspreis"].astype(str).str.replace("n132.81", "132.81")
df_2023_cleaned["Einkaufpreis"] = df_2023_cleaned["Einkaufpreis"].astype(str).str.replace("n96.71", "96.71")

df_2023_cleaned["Verkaufspreis"] = pd.to_numeric(df_2023_cleaned["Verkaufspreis"], errors="coerce")
df_2023_cleaned["Einkaufpreis"] = pd.to_numeric(df_2023_cleaned["Einkaufpreis"], errors="coerce")
df_2023_cleaned.head()

,Unnamed: 0,Getränk,Region,Bestellnummer,Rechnungsnummer,Kundennummer,Listenpreis,Rabatt in Prozent,Verkaufspreis,Rabatt Preis,Einkaufpreis,VK Peris in USD
0,205,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,8920,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,4116,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,1716,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,10901,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46


**FORTGESCHRITTEN** 
- löse das Problem aus der Zelle vorher nun jeweils generisch
- sodass die Buchstaben in den beiden Columns `Verkaufspreis` und `Einkaufpreis` entfernen kannst
<br><br>
**Tipp** 
- verwende u.a. regex
    - r'\D' funktioniert nicht, weil es nur Zahlen zurücklässt 
    - '\d+' funktioniert nicht, weil nur Zahlen gelöscht werden

In [1307]:
df_2023_cleaned = df_2023.copy()

price_cols = ["Verkaufspreis", "Einkaufpreis"]

df_2023_cleaned[price_cols] = (
    df_2023_cleaned[price_cols]
        .astype(str)
        .replace(r"[^0-9.\-]", "", regex=True)
        .apply(pd.to_numeric, errors="coerce")
)

df_2023_cleaned

,Unnamed: 0,Getränk,Region,Bestellnummer,Rechnungsnummer,Kundennummer,Listenpreis,Rabatt in Prozent,Verkaufspreis,Rabatt Preis,Einkaufpreis,VK Peris in USD
0,205,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,8920,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,4116,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,1716,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,10901,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...,...
10922,5585,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,6251,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,10719,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,6407,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


Verändere alle Columnnamen. Es sollen die Colummnamen wie folgt angepasst werden: 
- Namen bestehen nur aus `Grossbuchtaben`
- Leerzeichen werden durch `_` ersetzt
- Sonderzeichen sollen angepasst werden in z.B. ö zu oe usw.
- die Column `Unnamed: 0` wird gelöscht
<br><br> rows 10.927

In [1308]:
#for col in df_2023_cleaned.columns:
#    df_2023_cleaned[col] = df_2023_cleaned[col].astype(str)

df_2023_cleaned.drop("Unnamed: 0", axis=1, inplace=True)
df_2023_cleaned.columns = (df_2023_cleaned.columns.str.upper()
                           .str.replace(" ", "_")
                           .str.replace("Ä", "AE")
                           .str.replace("Ö", "OE")
                           .str.replace("Ü", "UE")
                           .str.replace("ß", "SS")
)

df_2023_cleaned

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- entferne die möglichen Doubletten in den Daten
<br><br> rows 10.303

In [1309]:
df_2023_cleaned.drop_duplicates(inplace=True)
df_2023_cleaned

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- entferne alle `null` rows
- also Einträge auf row Ebene, welchen keinen Werte beinhalten -> in der gesamten Reihe
<br><br> rows 10.302

In [1310]:
df_2023_cleaned = df_2023_cleaned.dropna(how="all")
df_2023_cleaned

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- Gibt es in den Daten Einträge ohne Rechnungsnummer? 
    - Wenn ja wie viele Einträge sind es? 

In [1311]:
df_2023_cleaned["RECHNUNGSNUMMER"].isna().value_counts()

RECHNUNGSNUMMER
False    10294
True         8
Name: count, dtype: int64

- Können Einträge ohne Rechnungsnummer nachträglich logisch abgeleitet werden aus den Daten
    - wenn ja bitte befülle die Rechnungsnummerngebe die `NaN` Werte in einem eigenen Dataframe an, 
    - um die Frage der Befüllung der Rechnungsnummer besser zu beantworten gebe die `NaN` Werte in einem eigenen Dataframe an

In [1312]:
#????? FRAGESTELLUNG ?????
df_rn = df_2023_cleaned[df_2023_cleaned["RECHNUNGSNUMMER"].isna()].copy()
df_rn

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
2171,Whisky,NaN,10078.0,NaN,54.0,NaN,5.0,NaN,-4.82,NaN,98.56
2265,Whisky,Asien,10004.0,NaN,54.0,148.21,5.0,148.21,-7.41,111.16,151.44
2684,Whisky,USA,10026.0,NaN,19.0,82.51,NaN,82.51,-16.50,61.88,71.00
6205,Rum,USA,10074.0,NaN,27.0,131.09,0.0,131.09,0.00,98.32,141.00
6332,Whisky,EU,10039.0,NaN,24.0,150.72,0.0,150.72,0.00,NaN,162.11
6351,NaN,EU,NaN,NaN,49.0,123.67,0.0,123.67,0.00,NaN,133.02
6886,Whisky,EU,10008.0,NaN,42.0,175.61,0.0,175.61,0.00,131.71,188.89
7890,Whisky,EU,10050.0,NaN,42.0,107.43,NaN,107.43,-21.49,80.57,92.44


- Was ist die Logik für die Wertbefüllung der Column `BESTELLNUMMER` und `RECHNUNGSNUMMER`

In [1313]:
"""
Bestellnummer: ganzzahlig iterierender float??
Rechnungsnummer: RE_{JAHR}_{BESTELLNUMMER}
"""

'\nBestellnummer: ganzzahlig iterierender float??\nRechnungsnummer: RE_{JAHR}_{BESTELLNUMMER}\n'

- Daten die sowohl einen `NaN` Wert bei der `BESTELLNUMMER` als auch bei der `RECHNUNGSNUMMER` sollen aud dem Datensatz entfernt werden
<br><br> rows 10.301

In [1314]:
df_2023_cleaned = df_2023_cleaned.dropna(
    subset=["BESTELLNUMMER", "RECHNUNGSNUMMER"],
    how="all"
)

df_2023_cleaned

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- Befüllen der NaN Werte in der Column `Rechnungsnummer`

In [1315]:
df_2023_cleaned["RECHNUNGSNUMMER"] = (
    df_2023_cleaned["RECHNUNGSNUMMER"]
    .fillna("RE_2023_" + df_2023_cleaned["BESTELLNUMMER"].astype(str))
)

df_2023_cleaned

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- Befüllen der NaN Werte in `Bestellnummer`
<br> `TIPP:` apply und lambda

In [1316]:
df_2023_cleaned["BESTELLNUMMER"] = (df_2023_cleaned["BESTELLNUMMER"]
                                    .fillna(df_2023_cleaned["RECHNUNGSNUMMER"].astype(str)
                                    .str.split("_").str[2]).astype(float))
df_2023_cleaned
#df_bn = df_2023_cleaned[df_2023_cleaned["BESTELLNUMMER"].isna()].copy()
#df_bn.head(10)

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


`Rechnungsnummern müssen einmalig sein`
- Aus diesem Grund sollen jeweils alle gleichen Rechnungsnummern aus dem Datensatz entfernt werden
- Erstelle hierfür einen neuen Dataframe `data_2023_unique`
<br><br> rows 10.079

In [1317]:
data_2023_unique = df_2023_cleaned.drop_duplicates(subset="RECHNUNGSNUMMER")
data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- befülle die NaN Werte in `RABATT_IN_PROZENT` mit den richtigen Prozenten
    - Berechnung ergibt sich aus `LISTENPREIS` und `VERKAUFSPREIS`

In [1318]:
data_2023_unique["RABATT_IN_PROZENT"] = data_2023_unique["RABATT_IN_PROZENT"].fillna((data_2023_unique["LISTENPREIS"] - data_2023_unique["VERKAUFSPREIS"]) / data_2023_unique["LISTENPREIS"]  * 100).astype(float)

data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- befülle die restlichen Werte von `RABATT_IN_PROZENT` mit der Zahl `0`

In [1319]:
data_2023_unique["RABATT_IN_PROZENT"] = data_2023_unique["RABATT_IN_PROZENT"].fillna(0.0).astype(float)
data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- befülle die NaN Werte in der Column `LISTENPREIS` mit dem Mittelwert der Column

In [1320]:
data_2023_unique["LISTENPREIS"] = data_2023_unique["LISTENPREIS"].fillna(data_2023_unique["LISTENPREIS"].median())
data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- befülle die NaN Werte mit dem `vorherigen` nicht NaN-Wert in der Column `VERKAUFSPREIS` --> forward fill verwenden

In [1321]:
data_2023_unique["VERKAUFSPREIS"] = data_2023_unique["VERKAUFSPREIS"].ffill()
data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- befülle die NaN Werte mit dem `nachstehenden` nicht NaN-Wert in der Column `EINKAUFPREIS` --> backward fill verwenden

In [1322]:
data_2023_unique["EINKAUFPREIS"] = data_2023_unique["EINKAUFPREIS"].bfill()
data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- schaue dir die Daten mit der help-function `info` an

In [1323]:
data_2023_unique.info()

<class 'pandas.DataFrame'>
Index: 10079 entries, 0 to 10926
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   GETRAENK           10066 non-null  str    
 1   REGION             10077 non-null  str    
 2   BESTELLNUMMER      10079 non-null  float64
 3   RECHNUNGSNUMMER    10079 non-null  str    
 4   KUNDENNUMMER       10079 non-null  float64
 5   LISTENPREIS        10079 non-null  float64
 6   RABATT_IN_PROZENT  10079 non-null  float64
 7   VERKAUFSPREIS      10079 non-null  float64
 8   RABATT_PREIS       10079 non-null  float64
 9   EINKAUFPREIS       10079 non-null  float64
 10  VK_PERIS_IN_USD    10079 non-null  float64
dtypes: float64(8), str(3)
memory usage: 944.9 KB


### c) Datenbearbeitung Wiederholung
- setze einen Filter von dem Dataframe `data_2023_neu_unique`
- durch den Filter sollen `alle` Einträge deren `BESTELLNUMMER` zwischen `10` und `20` ist angezeigt werden
<br><br> rows 11

In [1324]:
data_2023_filtered = data_2023_unique.copy()

data_2023_filtered = data_2023_filtered[(data_2023_unique["BESTELLNUMMER"] >= 10) & (data_2023_unique["BESTELLNUMMER"] <= 20)]
data_2023_filtered

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
288,Whisky,Asien,15.0,RE_2023_15,137.0,138.20,0.0,138.20,0.00,103.65,148.65
1055,Whisky,EU,10.0,RE_2023_10,296.0,143.53,5.0,143.53,-7.18,107.65,146.66
2693,Wein,Asien,19.0,RE_2023_19,84.0,164.63,0.0,164.63,0.00,123.47,177.08
3307,Whisky,USA,11.0,RE_2023_11,278.0,213.64,5.0,213.64,-10.68,160.23,218.30
3461,Champagner,Asien,17.0,RE_2023_17,40.0,157.89,0.0,157.89,0.00,118.42,169.83
3783,Whisky,USA,20.0,RE_2023_20,287.0,199.37,0.0,199.37,0.00,149.53,214.44
4132,Whisky,EU,18.0,RE_2023_18,87.0,131.38,0.0,131.38,0.00,98.54,141.31
7340,Champagner,EU,12.0,RE_2023_12,85.0,111.10,0.0,111.10,0.00,83.32,119.50
7966,Wein,EU,14.0,RE_2023_14,279.0,88.18,0.0,88.18,0.00,66.14,94.85
9319,Whisky,EU,16.0,RE_2023_16,169.0,100.78,0.0,100.78,0.00,75.59,108.40


- entferne alle rows mit den Kundennummern aus der Aufgabe vorher mittels `isin`
<br><br> rows 9.720

In [1325]:
customer_list = data_2023_filtered["KUNDENNUMMER"]
data_2023_unique = data_2023_unique[~data_2023_unique["KUNDENNUMMER"].isin(customer_list)]
data_2023_unique

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48


- sortiere die Daten einmal nach der `KUNDENNUMMER`

In [1326]:
data_2023_unique.sort_values("KUNDENNUMMER")

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
53,Champagner,USA,6312.0,RE_2023_6312,1.0,166.26,0.0,166.26,0.00,124.70,178.83
5563,Wein,EU,2841.0,RE_2023_2841,1.0,191.18,0.0,191.18,0.00,143.38,205.63
3128,Whisky,EU,1280.0,RE_2023_1280,1.0,186.82,0.0,186.82,0.00,140.12,200.94
6617,Whisky,EU,1079.0,RE_2023_1079,1.0,152.77,0.0,152.77,0.00,114.58,164.32
1717,Wein,Asien,2496.0,RE_2023_2496,1.0,80.87,0.0,80.87,0.00,60.65,86.98
...,...,...,...,...,...,...,...,...,...,...,...
3031,Whisky,Asien,7524.0,RE_2023_7524,299.0,108.93,10.0,108.93,-10.89,81.70,105.45
3901,Whisky,USA,2073.0,RE_2023_2073,299.0,205.34,0.0,205.34,0.00,154.00,220.86
11,Whisky,EU,9231.0,RE_2023_9231,299.0,228.97,5.0,228.97,-11.45,171.73,233.96
8903,Rum,EU,2049.0,RE_2023_2049,299.0,229.39,0.0,229.39,0.00,172.04,246.73


- sortiere die Daten nach zwei Kriterien (das erste ist `GETRAENK`und das zweite ist `BESTELLNUMMER`) 
    - `GETRAENK` wird **absteigend** und 
    . `BESTELLNUMMER` wird **aufsteigend** sortiert

In [1327]:
data_2023_unique.sort_values(by=["GETRAENK", "BESTELLNUMMER"], ascending=[False, True])

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
6224,Whisky,EU,0.0,RE_2023_0,26.0,88.37,0.0,88.37,0.00,66.28,95.05
10638,Whisky,EU,1.0,RE_2023_1,86.0,208.99,10.0,208.99,-20.90,156.74,202.31
174,Whisky,USA,6.0,RE_2023_6,215.0,130.06,5.0,130.06,-6.50,97.54,132.90
8254,Whisky,Asien,9.0,RE_2023_9,286.0,111.78,10.0,111.78,-11.18,83.84,108.21
6139,Whisky,EU,21.0,RE_2023_21,104.0,130.35,0.0,130.35,0.00,97.76,140.20
...,...,...,...,...,...,...,...,...,...,...,...
1041,NaN,EU,10059.0,RE_2023_10059,15.0,168.15,5.0,168.15,-8.41,126.11,171.82
3001,NaN,EU,10062.0,RE_2023_10062,56.0,110.21,0.0,110.21,0.00,82.66,118.54
10772,NaN,EU,10064.0,RE_2023_10064,62.0,119.71,0.0,119.71,0.00,89.78,128.76
10007,NaN,USA,10066.0,RE_2023_10066,56.0,148.81,0.0,148.81,0.00,71.66,160.06


## Aufgabe 4: Datenzusammenführen
### a) führe die Daten `data_2020, data_2021, data_2022 **und** data_2023_neu_unique` zusammen

#### fasse die Daten sinnvoll zusammen (Vorbereitung)
- schreibe eine Funktion um die Datensätze in die gleiche Columns umzuwandeln und erstelle jeweile neue Datframes mit dem Titel:
    - data_2020_n
    - data_2021_n
    - data_2022_n
- Die Einzelteile der Funktion wurden in den vorherigen Aufgaben schon definiert

In [1328]:
#Gottlose Aufgabenstellung??????????????????

def normalize_df(df):
    new_df = df.copy()

    new_df.drop("Unnamed: 0", axis=1, inplace=True)
    new_df.columns = (new_df.columns.str.upper()
                      .str.replace(" ", "_")
                      .str.replace("Ä", "AE")
                      .str.replace("Ö", "OE")
                      .str.replace("Ü", "UE")
                      .str.replace("ß", "SS")
                      )
    return new_df

data_2020_n = normalize_df(df_2020)
data_2021_n = normalize_df(df_2021)
data_2022_n = normalize_df(df_2022)

data_2020_n.columns.tolist() == data_2021_n.columns.tolist() == data_2022_n.columns.tolist() == data_2023_unique.columns.tolist()

True

- benutze für das zusammenführen einmal `concat` und erstelle einen neuen Dataframes `data_all`
<br><br> Lösung *25079 rows × 11 columns*

In [1329]:
#ignore_index = True da sonst probleme mit der indexierung entsteht
data_all = pd.concat([data_2020_n, data_2021_n, data_2022_n, data_2023_unique], ignore_index=True)
data_all.shape


(24720, 11)

- benutze hierfür einmal `append`
<br><br> Lösung *25079 rows × 11 columns*

In [1330]:
#data_2020_n.append([data_2021_n, data_2022_n, data_2023_unique],ignore_index=True)
#Gibts in dieser pandas version wohl nicht mehr

# `**ACHTUNG** APPEND NORMALERWEISE NICHT VERWENDEN`
## append wird in zukünftigen Versionen nicht mehr funktionieren

- benutze für das zusammenführen `merge`
<br><br> Lösung *25079 rows × 11 columns*

In [1331]:
# das ergibt einfach keinen sinn
# man merged nur wenn man einen eindeutigen key hat (id) aber das haben wir hier nicht
# wirklich niemals jemand

temp_df_20_21 = pd.merge(data_2020_n, data_2021_n, how="outer")
temp_df_22_23 = pd.merge(data_2022_n, data_2023_unique, how="outer")
temp_df_20_23 = pd.merge(temp_df_20_21, temp_df_22_23, how="outer")

temp_df_20_23

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Champagner,Asien,3.0,RE_2021_3,69.0,179.72,0.0,179.72,0.00,134.79,193.31
1,Champagner,Asien,4.0,RE_2022_4,163.0,168.89,5.0,160.45,-8.44,126.67,172.58
2,Champagner,Asien,7.0,RE_2020_7,97.0,202.31,0.0,202.31,0.00,151.73,217.60
3,Champagner,Asien,33.0,RE_2022_33,182.0,82.64,5.0,78.51,-4.13,61.98,84.45
4,Champagner,Asien,70.0,RE_2021_70,30.0,229.91,5.0,218.41,-11.50,172.43,234.92
...,...,...,...,...,...,...,...,...,...,...,...
24715,NaN,USA,10015.0,RE_2023_10015,58.0,217.54,10.0,217.54,-21.75,86.76,210.59
24716,NaN,USA,10017.0,RE_2023_10017,61.0,184.30,0.0,184.30,0.00,138.23,198.23
24717,NaN,USA,10048.0,RE_2023_10048,5.0,119.00,0.0,119.00,0.00,106.64,128.00
24718,NaN,USA,10058.0,RE_2023_10058,23.0,84.52,0.0,84.52,0.00,63.39,90.91


- benutze für das zusammenführen der Daten `join`
- funktioniert das?

In [1332]:
#ergibt noch weniger sinn weil man automatisch über den index joined => sorgt dafür, dass die einträge mit den selben indexen einfach überschrieben werden
data_2020_n.join(data_2021_n, how="outer", lsuffix="_2020", rsuffix="_2021")

,GETRAENK_2020,REGION_2020,BESTELLNUMMER_2020,RECHNUNGSNUMMER_2020,KUNDENNUMMER_2020,LISTENPREIS_2020,RABATT_IN_PROZENT_2020,VERKAUFSPREIS_2020,RABATT_PREIS_2020,EINKAUFPREIS_2020,...,REGION_2021,BESTELLNUMMER_2021,RECHNUNGSNUMMER_2021,KUNDENNUMMER_2021,LISTENPREIS_2021,RABATT_IN_PROZENT_2021,VERKAUFSPREIS_2021,RABATT_PREIS_2021,EINKAUFPREIS_2021,VK_PERIS_IN_USD_2021
0,Rum,Asien,0.0,RE_2020_0,27.0,86.23,0.0,86.23,0.00,64.67,...,EU,0,RE_2021_0,114,210.38,0,210.38,0.00,157.78,226.28
1,Wein,EU,1.0,RE_2020_1,87.0,182.93,0.0,182.93,0.00,137.20,...,Asien,1,RE_2021_1,57,223.11,0,223.11,0.00,167.33,239.98
2,Wein,Asien,2.0,RE_2020_2,76.0,103.97,0.0,103.97,0.00,77.98,...,EU,2,RE_2021_2,125,218.12,0,218.12,0.00,163.59,234.61
3,Whisky,USA,3.0,RE_2020_3,44.0,119.95,0.0,119.95,0.00,89.96,...,Asien,3,RE_2021_3,69,179.72,0,179.72,0.00,134.79,193.31
4,Whisky,EU,4.0,RE_2020_4,25.0,217.43,5.0,206.56,-10.87,163.07,...,EU,4,RE_2021_4,71,83.83,5,79.64,-4.19,62.87,85.66
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,EU,4995,RE_2021_4995,128,126.20,5,119.89,-6.31,94.65,128.95
4996,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,EU,4996,RE_2021_4996,130,130.73,0,130.73,0.00,98.05,140.61
4997,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,USA,4997,RE_2021_4997,103,221.26,0,221.26,0.00,165.94,237.99
4998,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,EU,4998,RE_2021_4998,141,112.46,0,112.46,0.00,84.34,120.96


### b) Newsletter Daten
- importiere die Dataframes `news` von den Jahren 2020 bis 2023 jeweils in einer eigenen Variabel

In [1333]:
news_origin_2020 = pd.read_excel("ressources/news_2020.xlsx")
news_origin_2021 = pd.read_csv("ressources/news_2021.csv", sep="\t")
news_origin_2022 = pd.read_csv("ressources/news_2022.csv", sep="\t")
news_origin_2023 = pd.read_csv("ressources/news_2023.csv", sep="\t")

print(news_origin_2020.columns)
print(news_origin_2021.columns)
print(news_origin_2022.columns)
print(news_origin_2023.columns)

Index(['Kd.Nr', 'Newsletter_1_Cupon_2020', 'Newsletter_2_Cupon_2020',
       'Newsletter_3_Cupon_2020', 'Newsletter_4_Cupon_2020',
       'Newsletter_5_Cupon_2020', 'Newsletter_6_Cupon_2020',
       'Newsletter_7_Cupon_2020', 'Newsletter_8_Cupon_2020',
       'Newsletter_9_Cupon_2020', 'Newsletter_10_Cupon_2020',
       'Newsletter_11_Cupon_2020', 'Newsletter_12_Cupon_2020'],
      dtype='str')
Index(['Unnamed: 0', 'Newsletter_1_Cupon_2021', 'Newsletter_2_Cupon_2021',
       'Newsletter_3_Cupon_2021', 'Newsletter_4_Cupon_2021',
       'Newsletter_5_Cupon_2021', 'Newsletter_6_Cupon_2021',
       'Newsletter_7_Cupon_2021', 'Newsletter_8_Cupon_2021',
       'Newsletter_9_Cupon_2021', 'Newsletter_10_Cupon_2021',
       'Newsletter_11_Cupon_2021', 'Newsletter_12_Cupon_2021', 'Nummer_Kunde'],
      dtype='str')
Index(['KD_NEWS_KEY', 'Newsletter_1_Cupon_2022', 'Newsletter_2_Cupon_2022',
       'Newsletter_3_Cupon_2022', 'Newsletter_4_Cupon_2022',
       'Newsletter_5_Cupon_2022', 'Newsletter_

- verändere den Index im Dataframe `news_2020` 
- der neue Index soll die `Kd.Nr` sein

In [1334]:
test = news_origin_2020.set_index(keys="Kd.Nr", inplace=False, drop=True).copy()
test

,Newsletter_1_Cupon_2020,Newsletter_2_Cupon_2020,Newsletter_3_Cupon_2020,Newsletter_4_Cupon_2020,Newsletter_5_Cupon_2020,Newsletter_6_Cupon_2020,Newsletter_7_Cupon_2020,Newsletter_8_Cupon_2020,Newsletter_9_Cupon_2020,Newsletter_10_Cupon_2020,Newsletter_11_Cupon_2020,Newsletter_12_Cupon_2020
Kd.Nr,,,,,,,,,,,,
33,nein,ja,nein,nein,ja,ja,nein,nein,nein,nein,ja,ja
91,ja,nein,nein,nein,nein,ja,ja,ja,nein,nein,nein,nein
23,nein,ja,ja,ja,ja,ja,nein,nein,nein,nein,nein,nein
16,nein,nein,nein,nein,nein,ja,ja,nein,nein,nein,nein,nein
19,nein,ja,nein,nein,nein,ja,ja,nein,nein,nein,nein,ja
70,nein,nein,nein,nein,nein,ja,ja,ja,nein,nein,nein,ja
51,nein,nein,ja,nein,nein,nein,ja,nein,nein,nein,nein,nein
67,nein,nein,nein,nein,nein,ja,ja,ja,nein,ja,ja,nein
3,nein,nein,ja,nein,ja,ja,ja,ja,nein,nein,nein,nein


- führe die Newsletter Dataframe aus den Jahren `2020, 2021, 2022` zusammen
- gebe nur die Kunden wieder, welche in allen 3 Datensätzen enthalten sind
<br><br> Lösung *24 rows × 38 columns*

In [1335]:
newsletter_all = (
    news_origin_2020
    .merge(news_origin_2021, how="inner", left_on="Kd.Nr", right_on="Unnamed: 0")
    .merge(news_origin_2022, how="inner", left_on="Kd.Nr", right_on="KD_NEWS_KEY")
)
newsletter_all.shape

# Habe mehrere Lösungswege probiert, kp wie er auf 24 Einträge kommt, ich komme immer auf 21

(24, 40)

- führe die Newsletter Dataframes aus den Jahren `2023, 2022` zusammen
- spiele dabei alle Werte von 2022 an den Dataframe von 2023 dran
- verwende `merge` für das zusammen führen
<br><br> Lösung *240 rows × 25 columns*

In [1336]:
#sollten 26 columns sein aber ist an sich für diese aufgabe eh der irrelevante teil
#abgesehen davon ist die aufgabe cursed -> es gibt 2 schein-id`s die sich immer um +1 unterscheiden
newsletter_merged = news_origin_2023.merge(
    news_origin_2022,
    on="KD_NEWS_KEY",
    how="left"
)
newsletter_merged

,Unnamed: 0,Newsletter_1_Cupon_2023,Newsletter_2_Cupon_2023,Newsletter_3_Cupon_2023,Newsletter_4_Cupon_2023,Newsletter_5_Cupon_2023,Newsletter_6_Cupon_2023,Newsletter_7_Cupon_2023,Newsletter_8_Cupon_2023,Newsletter_9_Cupon_2023,...,Newsletter_3_Cupon_2022,Newsletter_4_Cupon_2022,Newsletter_5_Cupon_2022,Newsletter_6_Cupon_2022,Newsletter_7_Cupon_2022,Newsletter_8_Cupon_2022,Newsletter_9_Cupon_2022,Newsletter_10_Cupon_2022,Newsletter_11_Cupon_2022,Newsletter_12_Cupon_2022
0,42,nein,nein,nein,nein,nein,ja,nein,nein,nein,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,nein,nein,nein,nein,ja,ja,nein,ja,nein,...,ja,nein,nein,nein,ja,nein,nein,nein,nein,nein
2,188,nein,nein,nein,nein,nein,nein,ja,ja,nein,...,nein,nein,nein,ja,nein,nein,nein,nein,ja,ja
3,216,ja,nein,nein,nein,ja,nein,nein,ja,nein,...,nein,ja,nein,ja,ja,nein,nein,ja,nein,nein
4,231,nein,nein,nein,nein,nein,nein,nein,nein,nein,...,nein,ja,nein,ja,nein,nein,nein,nein,nein,ja
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
235,254,nein,nein,ja,nein,nein,ja,nein,nein,nein,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
236,201,nein,nein,nein,nein,nein,ja,ja,nein,nein,...,nein,ja,nein,ja,nein,nein,ja,nein,nein,nein
237,165,nein,nein,nein,nein,ja,ja,ja,nein,ja,...,nein,nein,ja,nein,ja,ja,nein,nein,nein,nein
238,4,nein,nein,nein,nein,ja,nein,ja,ja,nein,...,ja,nein,ja,nein,ja,nein,nein,nein,nein,ja


- schreibe ***alle*** `news_data` in einen neuen Dataframes `news_all_years`
    - was kann ich verwenden, wenn ich `nicht merge` verwenden möchte. Begründe es kurz
- gebe nur eine `Kundennummer` in dem Dataframe als Columnnamen an
- **Tipp** reset index bei dem Dataframe `news_2020`
<br><br> Ergebnis 284 rows × 49 columns

In [1337]:
news_origin_2020 = news_origin_2020.rename(columns={"Kd.Nr": "Kd.Nr"})
news_origin_2021 = news_origin_2021.rename(columns={"Unnamed: 0": "Kd.Nr"})
news_origin_2022 = news_origin_2022.rename(columns={"KD_NEWS_KEY": "Kd.Nr"})
news_origin_2023 = news_origin_2023.rename(columns={"Unnamed: 0": "Kd.Nr"})

news_origin_2021.drop(columns="Nummer_Kunde", inplace=True)
news_origin_2023.drop(columns="KD_NEWS_KEY", inplace=True)

news_all_years = (news_origin_2020
                  .merge(news_origin_2021, how="outer", on="Kd.Nr")
                  .merge(news_origin_2022, how="outer", on="Kd.Nr")
                  .merge(news_origin_2023, how="outer", on="Kd.Nr"))


news_all_years

,Kd.Nr,Newsletter_1_Cupon_2020,Newsletter_2_Cupon_2020,Newsletter_3_Cupon_2020,Newsletter_4_Cupon_2020,Newsletter_5_Cupon_2020,Newsletter_6_Cupon_2020,Newsletter_7_Cupon_2020,Newsletter_8_Cupon_2020,Newsletter_9_Cupon_2020,...,Newsletter_3_Cupon_2023,Newsletter_4_Cupon_2023,Newsletter_5_Cupon_2023,Newsletter_6_Cupon_2023,Newsletter_7_Cupon_2023,Newsletter_8_Cupon_2023,Newsletter_9_Cupon_2023,Newsletter_10_Cupon_2023,Newsletter_11_Cupon_2023,Newsletter_12_Cupon_2023
0,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1,ja,ja,nein,nein,ja,nein,ja,nein,nein,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,nein,nein,ja,ja,nein,ja,nein,nein,nein,nein
3,3,nein,nein,ja,nein,ja,ja,ja,ja,nein,...,ja,nein,nein,ja,ja,ja,nein,nein,ja,nein
4,4,nein,nein,ja,nein,ja,nein,ja,ja,nein,...,nein,nein,ja,nein,ja,ja,nein,nein,nein,nein
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
275,295,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,nein,nein,nein,nein,nein,nein,nein,nein,ja,nein
276,296,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,ja,nein,ja,nein,ja,nein,nein,nein,nein,nein
277,297,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,nein,nein,nein,nein,nein,ja,nein,nein,nein,nein
278,298,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,nein,nein,ja,nein,ja,ja,nein,nein,ja,ja


- füge über die Kundennummern die Daten aus `news_all_years` an die Daten aus `data_all` an
- und speicher diese in dem Datensatz `data_merged_save`
<br><br> Ergebnis 25079 rows × 59 columns

In [1338]:
data_merged_save = pd.merge(data_all, news_all_years, how="outer", left_on="KUNDENNUMMER",right_on="Kd.Nr")
data_merged_save

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,...,Newsletter_3_Cupon_2023,Newsletter_4_Cupon_2023,Newsletter_5_Cupon_2023,Newsletter_6_Cupon_2023,Newsletter_7_Cupon_2023,Newsletter_8_Cupon_2023,Newsletter_9_Cupon_2023,Newsletter_10_Cupon_2023,Newsletter_11_Cupon_2023,Newsletter_12_Cupon_2023
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Whisky,USA,59.0,RE_2020_59,1.0,129.13,0.0,129.13,0.0,96.85,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Whisky,USA,62.0,RE_2020_62,1.0,218.32,0.0,218.32,0.0,163.74,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Wein,Asien,88.0,RE_2020_88,1.0,82.65,0.0,82.65,0.0,61.99,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Wein,EU,206.0,RE_2020_206,1.0,191.24,0.0,191.24,0.0,143.43,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24720,Rum,EU,2049.0,RE_2023_2049,299.0,229.39,0.0,229.39,0.0,172.04,...,ja,nein,nein,ja,nein,ja,nein,ja,nein,ja
24721,Champagner,USA,9894.0,RE_2023_9894,299.0,149.86,0.0,149.86,0.0,112.40,...,ja,nein,nein,ja,nein,ja,nein,ja,nein,ja
24722,Wein,EU,9703.0,RE_2023_9703,299.0,137.71,0.0,137.71,0.0,103.28,...,ja,nein,nein,ja,nein,ja,nein,ja,nein,ja
24723,Whisky,Asien,98.0,RE_2023_98,299.0,102.48,0.0,102.48,0.0,76.86,...,ja,nein,nein,ja,nein,ja,nein,ja,nein,ja


## Aufgabe 5: Daten in eine CSV und Excel schreiben

- speichere den Dataframe `data_merged_save` in einer Excel Datei
    - erstelle hierfür einen Ordner in der du die Excel speichern willst
    - gebe dem Excel den Sheet Namen `news_2023` mit

In [1339]:
data_merged_save.to_excel("output/merged_data.xlsx", index=False, sheet_name="news_2023")

- speichere den Dataframe `data_merged_save` äquivalent in einer csv Datei
- gebe den `Seperator` und das `Encoding` verfahren mit

In [1340]:
data_merged_save.to_csv("output/merged_data.csv", index=False, encoding="utf-8", sep="\t")

## Aufgabe 6: Datenselektion
### a) selektion mittels Datenzusammenführen
- erstelle `nur` einen Dataframe von allen Kunden aus den Daten `data_2020`, welche auch Produkte in den anderen `data_2021, data_2022 und data_2023_neu_unique` haben
<br><br> Ergebnis 11.653 rows × 11 columns

In [1360]:
# concat # Ich verstehe die aufgabe nicht
# Ich erstelle ein Dataframe mit allen Kundennummern aus 2021-2023
kunden_pool = (
    pd.concat(
        [
            data_2021_n[["KUNDENNUMMER"]],
            data_2022_n[["KUNDENNUMMER"]],
            data_2023_unique[["KUNDENNUMMER"]],
        ],
        ignore_index=True
    )
    .drop_duplicates()
)
kunden_pool

,KUNDENNUMMER
0,114.0
1,57.0
2,125.0
3,69.0
4,71.0
...,...
13604,263.0
13607,254.0
13633,286.0
13790,292.0


In [1363]:
#merge
merge_2020 = data_2020_n.merge(
    kunden_pool,
    on="KUNDENNUMMER",
    how="inner"
).drop_duplicates(subset="KUNDENNUMMER")

merge_2020

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Rum,Asien,0,RE_2020_0,27,86.23,0,86.23,0.00,64.67,92.75
1,Wein,EU,1,RE_2020_1,87,182.93,0,182.93,0.00,137.20,196.76
2,Wein,Asien,2,RE_2020_2,76,103.97,0,103.97,0.00,77.98,111.83
3,Whisky,USA,3,RE_2020_3,44,119.95,0,119.95,0.00,89.96,129.02
4,Whisky,EU,4,RE_2020_4,25,217.43,5,206.56,-10.87,163.07,222.18
...,...,...,...,...,...,...,...,...,...,...,...
340,Whisky,Asien,340,RE_2020_340,19,184.62,0,184.62,0.00,138.46,198.58
365,Wein,USA,365,RE_2020_365,40,172.02,0,172.02,0.00,129.02,185.02
419,Champagner,EU,419,RE_2020_419,64,161.94,0,161.94,0.00,121.46,174.18
445,Whisky,USA,445,RE_2020_445,12,178.76,0,178.76,0.00,134.07,192.27


### b) nach Selektionskriterien in Dataframes
- selktiere den Dataframe `data_all` nach den folgenden Kriterien
#### Der Dataframe soll endeweder:
- `nur` die `REGION EU oder Asien` und
- nur Kunden mit einer `KUNDENNUMMER grösser/gleich 50` und
- das Getränk darf `nicht gleich Champagner` sein und
- es wurde kein Rabatt gewärt (`= 0`)
<br><br> **oder** <br><br>
- die `BESTELLNUMMER` enthält die Zahl 1 an irgendeiner Stelle z.B. `(1, 10, 21 ..)`
<br><br> Ergebnis 15.772 rows × 11 columns

In [1368]:
filter_data_all = data_all[
    (
        (
            (data_all["REGION"].isin(["EU", "Asien"])) &
            (data_all["KUNDENNUMMER"] >= 50) &
            (data_all["GETRAENK"] != "Champagner") &
            (data_all["RABATT_IN_PROZENT"] == 0)
        )
        |
        (
            data_all["BESTELLNUMMER"]
            .astype(str)
            .str.contains("1", regex=False)
        )
    )
]

filter_data_all

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
1,Wein,EU,1.0,RE_2020_1,87.0,182.93,0.0,182.93,0.0,137.20,196.76
2,Wein,Asien,2.0,RE_2020_2,76.0,103.97,0.0,103.97,0.0,77.98,111.83
9,Whisky,Asien,9.0,RE_2020_9,85.0,229.50,0.0,229.50,0.0,172.12,246.85
10,Rum,EU,10.0,RE_2020_10,41.0,164.81,0.0,164.81,0.0,123.61,177.27
11,Rum,EU,11.0,RE_2020_11,82.0,161.34,0.0,161.34,0.0,121.00,173.54
...,...,...,...,...,...,...,...,...,...,...,...
24713,Rum,USA,2119.0,RE_2023_2119,31.0,163.87,0.0,163.87,0.0,122.90,176.26
24714,Whisky,EU,3067.0,RE_2023_3067,270.0,200.23,0.0,200.23,0.0,150.17,215.37
24715,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.0,61.00,87.49
24716,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.0,168.04,240.99


#### selktiere den Dataframe `data_2023`
- es sollen nur rows angezeigt werden, welche
    - `kein Wert` in der Rechnungsnummer haben **`und`**
    - `ein Wert` in der Kundennummer haben
<br><br> Ergebnis 8 rows

In [1371]:
filter_data_2023 = data_2023_unique[
    (
        data_2023_unique["RECHNUNGSNUMMER"].isna()
    )
    &
    (
        ~data_2023_unique["KUNDENNUMMER"].isna()
    )
]
filter_data_2023

# codelogik ist korrekt aber ergebnis weicht mal wieder ab

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD


- es sollen nur rows angezeigt werden, welche
    - `ein Wert` in der Rabattcolumn haben **`oder`** 
    - `kein Wert` in der Verkaufspreis
<br><br> Ergebnis 10.920 rows × 12 columns

In [1373]:
filter_data_2023 = data_2023_unique[
    (
        ~data_2023_unique["RABATT_IN_PROZENT"].isna()
    )
    |
    (
        data_2023_unique["VERKAUFSPREIS"].isna()
    )
]

filter_data_2023

,GETRAENK,REGION,BESTELLNUMMER,RECHNUNGSNUMMER,KUNDENNUMMER,LISTENPREIS,RABATT_IN_PROZENT,VERKAUFSPREIS,RABATT_PREIS,EINKAUFPREIS,VK_PERIS_IN_USD
0,Whisky,USA,5645.0,RE_2023_5645,16.0,128.95,5.0,128.95,-6.01,96.71,131.76
1,Whisky,EU,7994.0,RE_2023_7994,193.0,132.81,5.0,132.81,-6.64,99.61,135.71
2,Whisky,USA,3190.0,RE_2023_3190,126.0,216.21,0.0,216.21,0.00,162.16,232.56
3,Whisky,USA,790.0,RE_2023_790,171.0,172.13,0.0,172.13,0.00,129.10,185.14
4,Rum,USA,9975.0,RE_2023_9975,41.0,128.73,0.0,128.73,0.00,96.55,138.46
...,...,...,...,...,...,...,...,...,...,...,...
10922,Whisky,Asien,4659.0,RE_2023_4659,267.0,81.34,0.0,81.34,0.00,61.00,87.49
10923,Wein,EU,5325.0,RE_2023_5325,253.0,224.05,0.0,224.05,0.00,168.04,240.99
10924,Whisky,USA,9793.0,RE_2023_9793,152.0,149.56,0.0,149.56,0.00,112.17,160.87
10925,Wein,EU,5481.0,RE_2023_5481,19.0,213.35,0.0,213.35,0.00,160.01,229.48
